zebra/zebra_srv6.c

```diff
@@ -0,0 +1,331 @@
+ */
+
+#include <zebra.h>
+
+#include "network.h"
+#include "prefix.h"
+#include "stream.h"
+#include "srv6.h"
+#include "zebra/zapi_msg.h"
+#include "zebra/zserv.h"
+#include "zebra/zebra_router.h"
+#include "zebra/zebra_srv6.h"
+#include "zebra/zebra_memory.h"
+#include "zebra/zebra_errors.h"
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <arpa/inet.h>
+#include <netinet/in.h>
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <arpa/inet.h>
+#include <netinet/in.h>
+
+
+static void set_seg6local_end(struct prefix_ipv6 *prefix)
+{
+	struct route_entry *re;
+	struct nexthop_group *ng;
+	struct vrf *vrf;
+	struct interface *ifp;
+	struct nexthop *nexthop;
+	int ret;
+	ifindex_t ifindex;
+
+	re = XCALLOC(MTYPE_RE, sizeof(struct route_entry));
+	re->type = ZEBRA_ROUTE_STATIC;
+	re->instance = 0;
+	re->uptime = monotime(NULL);
+	re->vrf_id = VRF_DEFAULT;
+	re->table = RT_TABLE_MAIN;
+	SET_FLAG(re->flags, ZEBRA_FLAG_SEG6LOCAL_ROUTE);
+
+	ifindex = 1;
+	vrf = vrf_lookup_by_id(VRF_DEFAULT);
+	FOR_ALL_INTERFACES (vrf, ifp) {
+		if (ifp->ifindex == 1)
+			continue;
+		if (!if_is_operative(ifp))
+			continue;
+		ifindex = ifp->ifindex;
+		break;
+	}
+
+	if (ifindex == 1)
+		zlog_err("%s: no available interface on default-vrf",
+			 __func__);
+
+	ng = nexthop_group_new();
+	nexthop = nexthop_from_ifindex(ifindex, VRF_DEFAULT);
+	if (!nexthop) {
+		flog_warn(
+			EC_ZEBRA_NEXTHOP_CREATION_FAILED,
+			"%s: Nexthops Specified: but we failed to properly create one",
+			__PRETTY_FUNCTION__);
+		nexthop_group_delete(&ng);
+		XFREE(MTYPE_RE, re);
+		return;
+	}
+
+	nexthop_add_seg6local(nexthop, ZEBRA_SEG6_LOCAL_ACTION_END, NULL);
+	nexthop_group_add_sorted(ng, nexthop);
+
+	ret = rib_add_multipath(AFI_IP6, SAFI_UNICAST,
+				(struct prefix *)prefix, NULL, re, ng);
+	if (ret < 0)
+		zlog_err("%s: can't add RE entry to rib", __func__);
+}
+
+static void locator_init(const struct srv6_locator *loc)
+{
+	size_t idx_hi, idx_lo;
+	uint16_t value = 0x0001;
+	struct prefix_ipv6 p;
+	struct srv6_function *function;
+
+	idx_hi = loc->prefix.prefixlen / 8;
+	idx_lo = idx_hi + 1;
+
+	p = loc->prefix;
+	p.prefix.s6_addr[idx_hi] = (0xff00 & value) >> 8;
+	p.prefix.s6_addr[idx_lo] = (0x00ff & value);
+	p.prefixlen = loc->prefix.prefixlen + loc->function_bits_length;
+	set_seg6local_end(&p);
+
+	function = srv6_function_alloc(&p);
+	listnode_add(loc->functions, function);
+	strlcpy(function->locator_name, loc->name,
+		sizeof(function->locator_name));
+}
+
+static void get_function_addr(struct in6_addr *locator,
+			      uint64_t current, int lshift)
+{
+	uint32_t *p = (uint32_t *)locator;
+	uint32_t tmp[4] = {};
+
+	if (lshift > 0x5f)
+		return;
+
+	tmp[lshift >> 5] = current << (lshift & 0x1f);
+	tmp[(lshift >> 5) + 1] = current >> (32 - (lshift & 0x1f));
+	p[0] |= ntohl(tmp[3]);
+	p[1] |= ntohl(tmp[2]);
+	p[2] |= ntohl(tmp[1]);
+	p[3] |= ntohl(tmp[0]);
```

Comment:
what is going on here

Reply:
This line was deleted when I re-write SRv6 manager. :bow: